### DISCUSSION TITLE
Any advice on getting an AjvValidator working in NextJS?

### DISCUSSION THREAD

**Discussion by eedeebee** [josdejong/svelte-jsoneditor#291] [Q&A] [answered]
Thanks for this great work!  Every time I try, the validator seems to work but I end up with 

react-dom.development.js:16348 Uncaught Error: document is not defined

which I presume is something in Ajv not quite working right in node land.  I've got it in side a 'use client' page, but that doesn't seem to help.  Any thoughts?

**Comment by josdejong** [comment]
The Ajv library and the wrapper function `createAjvValidator` both do not use `document` in their source code. Other code of `svelte-jsoneditor` does. I expect that in your application, the whole bundle is loaded on the server side, whereas ideally it should treeshake and only load the `createAjvValidator` part. 

In a React project of mine I've found a workaround for that by importing `createAjvValidator` and some other util functions from the `svelte-jsoneditor` library (so it can do treeshaking), and import `JSONEditor` from `vanilla-jsoneditor`. It is an ugly workaround though. Help finding a better solution would be welcome.

**Comment by eedeebee** [comment, marked as answer]
Thank you!  Here's what I had: 
```
'use client';
import dynamic from 'next/dynamic'

const JSONEditorReact = dynamic(() => import('@/components/ui/JSONEditorReact'), { ssr: false })
import type { JSONContent, TextContent, Content, OnChangeStatus, Validator } from 'vanilla-jsoneditor'
import { Mode, createAjvValidator } from 'vanilla-jsoneditor';
```

and in my wrapper component (JSONEditorReact)

`import { JSONEditor, JSONEditorPropsOptional } from 'vanilla-jsoneditor'
`

 I do think you're right that for some reason, all of vanilla-jsoneditor is being included on the server side, but I'm really not sure why or what's triggering it.  The weird thing is that if I don't bring in **and** call createAjvValidator(), I don't see the issue.   

Would moving the use of createAjvValidator to my wrapper component help?  I tried it and it did!  I pass my schema into the component and call the utility in the wrapper and the issues is gone!  I dunno if this is working as designed or indicative of some bug in next land...

**Reply by josdejong** [reply]
Thanks for sharing your case.

Right now the library exports a single, bundled ESM file (usable in the browser). It can indeed well be that if we keep the files separate, that it solves this issue. That would be interesting to try out. So then the `vanilla-jsoneditor` package should by default export individual ESM files, and have a "browser" field in `package.json` pointing to the bundled file.

Anyone able to help try that out?
